### ISSUE TITLE
"Checking auto-linked files" takes a long time

### ISSUE THREAD

**Issue by asklar** [microsoft/react-native-windows#5304]
Can this experience be improved?

6>ΓêÜ Checking auto-linked files...
6>Success: No auto-linking changes necessary. (8081ms)

**Comment by NickGerleman** [comment]
Seems like we could feed the lockfile as an input into the target to not retrigger the check unless packages change.

**Comment by NickGerleman** [comment]
Marking as a regression since 0.62

**Comment by jonthysell** [comment]
Does this repro when using `run-windows` or only when using VS? IIRC the command is way faster when just running from the command line.

**Comment by NickGerleman** [comment]
IIRC I've seen `run-windows` do the check twice, where we run it in JS, then a second time when we start MSBuild.

**Comment by jonthysell** [comment]
@NickGerleman Yes, since autolinking modifies project/solution files, we can't run auto-linking during msbuild. So `run-windows` actually runs auto linking _before_ msbuild, then during msbuild we run a convenience check to verify that autolinked files are up-to-date and produce warnings.

https://microsoft.github.io/react-native-windows/docs/next/native-modules-autolinking#autolinking-process

**Comment by asklar** [comment]
> Does this repro when using `run-windows` or only when using VS? IIRC the command is way faster when just running from the command line.

@jonthysell  do we need to run it when running in VS?

**Comment by NickGerleman** [comment]
> @NickGerleman Yes, since autolinking modifies project/solution files, we can't run auto-linking during msbuild. So `run-windows` actually runs auto linking _before_ msbuild, then during msbuild we run a convenience check to verify that autolinked files are up-to-date and produce warnings.
> 
> https://microsoft.github.io/react-native-windows/docs/next/native-modules-autolinking#autolinking-process

I guess I meant to say that I've seen the same slowdown happening in the CLI.

To cut down on running the checks twice it seems like the CLI could pass the MSBuild prop to disable autolinking, since it will already have run it. The CLI already passes some custom props right now, so that should be pretty simple I think.

I know I was the one who kept on pushing for VS integration. I do think it's super useful, but also could understand turning it off by default if we don't have great perf, or have other limitations right now.

**Comment by jonthysell** [comment]
I don't know at what point and where Alex did his tests, I just checked both for playground and the sample apps, and I never see 8 seconds.

The most expensive thing running is the automatic run of `react-native config` at the beginning of (every) cli command, and that is wholly dependent on how many projects need to be scanned. With our multiple app/library project setups in our repo, this takes between ~1500-2000ms, but only happens once at the start.

So when the actual auto-linking runs during `run-windows` it only takes ~120ms because we just pass it the config that was calculated for `run-windows`.

However when you just run `autolink-windows` on its own, now that time includes having to call `react-native config`. So for our apps:

```
npx react-native autolink-windows --check --sln windows\playground.sln --proj windows\playground\Playground.vcxproj
 √ Checking auto-linked files...
Success: No auto-linking changes necessary. (2053ms)
```

```
npx react-native autolink-windows --check --sln windows\SampleApps.sln --proj windows\SampleAppCS\SampleAppCS.csproj
 √ Checking auto-linked files...
Success: No auto-linking changes necessary. (1898ms)
```

```
npx react-native autolink-windows --check --sln windows\SampleApps.sln --proj windows\SampleAppCpp\SampleAppCpp.vcxproj
 √ Checking auto-linked files...
Success: No auto-linking changes necessary. (1750ms)
```

The longest I'm seeing, is that when msbuild runs the check, I can see anywhere from ~3100-5100ms. I don't know why msbuild takes a 1-3 extra seconds.

I don't think the auto-linking code itself can be optimized much further, when I did profiling a while back the biggest expense is disk I/O, and I already added logic to minimize writing to disk.

> To cut down on running the checks twice it seems like the CLI could pass the MSBuild prop to disable autolinking, since it will already have run it. 

Yes, this seems like a good idea for the cli to save a couple seconds. If you use `run-windows --no-autolink` it will disable both the actual run and the msbuild check - so we can just always disable the msbuild check.

For anyone else who is really seeing a problem, we can use the already working solution of explicitly setting the property to false, which is what we do for our solutions where we know we're not relying on auto-linking.

Also, I think if I can get #5168 figured out, so the check only runs when say package.json changes, that should also reduce how often the check happens and improve perf further.
